### PR TITLE
Crypto implementation for shared model

### PR DESCRIPTION
**Pull request by luckychess**
## What is this pull request?
Implementation for shared model crypto.
   
## Why do you implement it? Why do we need this pull request?
Shared model does not contain crypto implementation now, only interfaces. Also this is an important step to crypto bindings.
  
## Details/Features
- Move all crypto and hash code to shared model
- Add BlobImpl class as implementation of Blob interface
- Inherit BlobImpl instead of Blob for public and private keys, Signed, Hash
- Replace Keypair interface with implementation
- Inherit Keypair from Primitive; add makeOldModel
- Introduce new Seed object for storing seeds
- Add Signer and Verifier as low-level crypto wrappers
- Add CryptoProvider as a wrapper for all crypto operations
- Add HashProvider as a wrapper for all low-level hashing operations
- Use real hashes in proto Amount and AddAssetQuantity

### CODE REVIEW

File: shared_model/cryptography/keypair.hpp [hunk 1]
```diff
@@ -18,33 +18,39 @@
...
 
+#include "common/types.hpp"
+
 namespace shared_model {
   namespace crypto {
     /**
      * Class for holding a keypair: public key and private key
      */
-    class Keypair : public interface::ModelPrimitive<Keypair> {
+    class Keypair : public interface::Primitive<Keypair, iroha::keypair_t> {
```
Comment (motxx) [inline shared_model/cryptography/keypair.hpp]: What is the reason for this change?

File: shared_model/cryptography/ed25519_sha3_impl/crypto_provider.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+#include "cryptography/ed25519_sha3_impl/signer.hpp"
+#include "cryptography/ed25519_sha3_impl/verifier.hpp"
+#include "cryptography/seed.hpp"
+
+namespace shared_model {
+  namespace crypto {
+    /**
+     * Wrapper class for signing-related stuff.
+     */
+    class CryptoProvider {
```
Comment (motxx) [inline shared_model/cryptography/ed25519_sha3_impl/crypto_provider.hpp]: Have we reached any consensus on the discussions that we have to use static functions vs free functions? Sorry, I don't know much about the situation...

File: shared_model/cryptography/keypair.hpp (hunk 1 above)
Comment (luckychess) [reply shared_model/cryptography/keypair.hpp]: Well, I changed it because we actually have an old "model" object for ```keypair``` - ```keypair_t``` is not a primitive type like ```blob<N>``` but a struct. So I decided to implement ```Primitive``` interface and ```makeOldModel``` to work with ```keypair``` more conveniently.

File: shared_model/cryptography/ed25519_sha3_impl/crypto_provider.hpp (hunk 2 above)
Comment (luckychess) [reply shared_model/cryptography/ed25519_sha3_impl/crypto_provider.hpp]: I think there are no real difference between them because both are hard to mock. I introduced a class here to make it easier.

**Comment by Warchant** [comment]
|`shared_model/cryptography/blob_impl.hpp:35`|
|---|
what is this function? move outside (make free function)

File: iroha-cli/interactive/impl/interactive_query_cli.cpp [hunk 3]
```diff
@@ -20,8 +20,8 @@
 #include <fstream>
 
 #include "client.hpp"
-#include "crypto/crypto.hpp"
-#include "crypto/hash.hpp"
+#include "cryptography/ed25519_sha3_impl/internal/ed25519_impl.hpp"
```
[the remark this answers is not in the document]
Comment (luckychess) [reply iroha-cli/interactive/impl/interactive_query_cli.cpp]: These changes (and similar ones in other files) were done just because of moving crypto from libs to shared_model. It looks like now we use ed25519 with sha-3 crypto directly - and it does but it is absolutely the same as before. It will be improved in separate PR.

File: irohad/validation/CMakeLists.txt [hunk 4]
```diff
@@ -20,7 +20,7 @@ add_library(stateful_validator
     )
 target_link_libraries(stateful_validator
     optional
-    ed25519
+    ed25519_sha3
```
[the remark this answers is not in the document]
Comment (luckychess) [reply irohad/validation/CMakeLists.txt]: I think better to link with a smaller shared model part if we need only crypto and nothing else - just to make resulting binary smaller.

File: shared_model/backend/protobuf/commands/proto_add_asset_quantity.hpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+#include "interfaces/common_objects/types.hpp"
+
+#include "backend/protobuf/common_objects/amount.hpp"
+#include "commands.pb.h"
+#include "interfaces/commands/add_asset_quantity.hpp"
+#include "utils/lazy_initializer.hpp"
+
+namespace shared_model {
+  namespace proto {
+    class AddAssetQuantity final : public interface::AddAssetQuantity {
```
[the remark this answers is not in the document]
Comment (luckychess) [reply shared_model/backend/protobuf/commands/proto_add_asset_quantity.hpp]: Well, I did a rebase to branch with shared model commands for Lazy. That's a side effect :(

File: shared_model/backend/protobuf/common_objects/amount.hpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_PROTO_AMOUNT_HPP
+#define IROHA_PROTO_AMOUNT_HPP
+
+#include <numeric>
+#include "cryptography/ed25519_sha3_impl/hash_provider.hpp"
```
[the remark this answers is not in the document]
Comment (luckychess) [reply shared_model/backend/protobuf/common_objects/amount.hpp]: Sure. Will be fixed soon as a separate task.

File: shared_model/cryptography/ed25519_sha3_impl/crypto_provider.hpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+     public:
+      /**
+       * Signs the message.
+       * @param blob - blob to sign
+       * @param keypair - keypair
+       * @return Signed object with signed data
+       */
+      Signed sign(const Blob &blob, const Keypair &keypair) const;
```
[the remark this answers is not in the document]
Comment (luckychess) [reply shared_model/cryptography/ed25519_sha3_impl/crypto_provider.hpp]: Low-level API of ed25519 library requires both keys for signing. Probably it could be reworked to deduce public key from a private one but I don't think this PR is a good place to do this (and I think we shouldn't change ed25519 internals as far as we absolutely have no other choice).
